freshmaker/pyxis.py

```diff
@@ -0,0 +1,253 @@
+                    # Check if the right format of version is used
+                    try:
+                        if version.parse(new_ver) > version.parse(old_ver):
+                            bundle_per_channel[channel] = bundle
+                    except version.InvalidVersion as e:
+                        log.warning("Other format than SemVer is used")
+                        log.warning(repr(e))
+                else:
+                    # Check if the right format of version is used
+                    try:
+                        version.parse(bundle.get('version'))
+                        bundle_per_channel[channel] = bundle
+                    except version.InvalidVersion as e:
+                        log.warning("Other format than SemVer is used")
+                        log.warning(repr(e))
+
+            ret[index_image] = bundle_per_channel
+        return ret
+
+    def get_digests_by_nvrs(self, nvrs, params={}):
+        """
+        Get images' digests(manifest_list_digest field) by their nvrs
+
+        :param list nvrs: list of nvrs of ContainerImages to query Pyxis
+        :param dict params: parameters to add to GET request
+        :return: set of digests
+        :rtype: set
+        """
+        include_fields = ['data.brew', 'data.repositories.manifest_list_digest']
+        request_params = copy.deepcopy(params)
+        request_params['include'] = ','.join(include_fields)
+        images = []
+        for nvr in nvrs:
+            images = self._pagination(f'images/nvr/{nvr}', request_params)
+
+        # get all manifest_list_digest of ContainerImages we got from Pyxis
+        digests = set()
+        for image in images:
+            if image.get('repositories') and \
+                    'manifest_list_digest' in image['repositories'][0]:
+                digests.add(image['repositories'][0]['manifest_list_digest'])
+
+        return digests
+
+    def filter_bundles_by_nvrs(self, nvrs, bundles):
+        """
+        Filter out bundles that don't have same digests in 'related_image' as
+        repositories.*.manifest_list_digest in images
+
+        :param list nvrs: nvrs to find images and filter bundles by them
+        :param list bundles: bundles to filter
+        :return: filtered list of bundles
+        :rtype: list
+        """
+        # get images by nvrs and then get their digests
+        image_digests = self.get_digests_by_nvrs(nvrs)
+
+        if not image_digests:
+            log.info(f"The are no digests for nvrs: {','.join(nvrs)}")
+            return []
+
+        # if bundle doesn't have any of digests of images, remove that bundle
+        for bundle in bundles:
+            for image in bundle.get('related_images', []):
+                # If we found same digest within images' digests, continue with the next bundle
+                if image.get('digest') in image_digests:
+                    break
+            else:
+                bundles.remove(bundle)
```

Comment:
When modifying a list during an iteration, it seems like strange things are happening:
```
>>> t = list(range(5))
>>> for x in t:
...     t.remove(x)
... 
>>> 
>>> t
[1, 3]
```

Additionally, calling remove on a list is quite inefficient since it is an [O(n)](https://wiki.python.org/moin/TimeComplexity) operation every time.

You can either use [filter](https://docs.python.org/3/library/functions.html#filter) or you can iterate through a copy of `bundles` in reverse order and remove elements by using the index on the actual `bundles`. The memory usage of copying `bundles` should be minimal since as long as you don't use `deepcopy`, it'll just be a list of pointers under the hood, pointing to dictionary objects.

Reply:
Maybe it's better to just create a new list with just accepted bundles? So it will be something like this:
```
       ret_bundles = []
        # if bundle doesn't have any of digests of images, remove that bundle
        for bundle in bundles:
            for image in bundle.get('related_images', []):
                # If we found same digest within images' digests, continue with the next bundle
                if image.get('digest') in image_digests:
                    ret_bundles.append(bundle)
                    break

        return ret_bundles
```

Reply:
Yes, that would b e fine as well.